src/harness/reference_models/pre_iap_filtering/pre_iap_util.py

```diff
@@ -0,0 +1,186 @@
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+import shapely.geometry as sgeo
+from reference_models.geo import vincenty
+from sas_test_harness import generateCbsdReferenceId
+from reference_models.geo import utils
+
+FSS_GWBL_PROTECTION_DISTANCE = 150
+
+
+def purgeOverlappingGrants(cbsds, frequency_range):
+  """Removes Grants from CBSDs that overlap with the given frequency range.
+
+  Args:
+  cbsds: List of CbsdData objects. Grants are removed from the input.
+  frequency_range: A dictionary with keys 'lowFrequency' and 'highFrequency'
+  """
+  for cbsd in cbsds:
+    if not cbsd['grants']:
+      continue
+    cbsd['grants'] = [grant for grant in cbsd['grants']
+                            if not checkForOverlappingGrants(grant, frequency_range)]
+
+def checkForOverlappingGrants(grant, frequency_range):
+  """Check if a grant overlaps with a given frequency range
+
+  If the lowFrequency or the highFrequency of the grant falls within the
+  low and high frequencies of the given range then the grant is considered
+  to be overlapping.
+  Args:
+    grant: A GrantData object dictionary defined in SAS-SAS spec
+    frequency_range: A dictionary with keys 'lowFrequency' and 'highFrequency'
+  Returns:
+    True if the grant overlaps else False
+  """
+  low_frequency_cbsd = grant['operationParam']['operationFrequencyRange']['lowFrequency']
+  high_frequency_cbsd = grant['operationParam']['operationFrequencyRange']['highFrequency']
+  low_frequency = frequency_range['lowFrequency']
+  high_frequency = frequency_range['highFrequency']
+  if (low_frequency_cbsd <= low_frequency < high_frequency_cbsd) or \
+     (low_frequency_cbsd < high_frequency <= high_frequency_cbsd):
+      return True
+  return False
+
+
+def getCbsdsWithinPolygon(cbsds, polygon):
+  """Get the list of all CBSDs that are present within the polygon or on the 
+  boundary of the Polygon
+
+  Args:
+    cbsds: List of CbsdData dictionaries as defined in the SAS-SAS specification.
+    polygon: A GeoJSON object containing polygon information.
+  Returns:
+    List of CBSDs lying within or on the boundary of the protectionn area.
+  """ 
+  cbsds_within_polygon = []
+  polygon = utils.ToShapely(polygon['features'][0]['geometry'])
+  for cbsd in cbsds:
+    cbsd_lat = cbsd['registrationRequest']['installationParam']['latitude']
+    cbsd_long = cbsd['registrationRequest']['installationParam']['longitude']
+    point = sgeo.Point(cbsd_lat, cbsd_long)
+
+    # If the CBSD is within the polygon and has grants then add it to the list
+    if (polygon.contains(point) or polygon.touches(point)) and cbsd['grants']:
```

Comment:
I am not 100% of the need for: touches().
I agree that without this, a CBSD exactly in the border will not be included in the list.
But in some way it is equivalent to when we do floating point comparison of type:  float_value <= ref_value
It introduces subtle floating point issue with compilers and implementations. Please check with Mo on your side who went over such issues within the prop models.

Usually it is better to reframe things to avoid the equality part of the check by testing  either:
  + float_value < ref_value
  + or float_value < ref_value + epsilon

The == plays the role of the 'touches' here.
So please think on this, if it makes sense to remove the "touches()"

Reply:
we are working on this with Mo and will get back to you soon.

Reply:
Ok, please continue your analysis, and we can change that on another PR.